shotover-proxy/src/transforms/redis/sink_single.rs

```diff
@@ -73,43 +86,230 @@ impl Transform for RedisSinkSingle {
+                // Notes on subscription responses
+                //
+                // There are 3 types of pubsub responses and the type is determined by the first value in the array:
+                // * `subscribe` - a response to a SUBSCRIBE, PSUBSCRIBE or SSUBSCRIBE request
+                // * `unsubscribe` - a response to an UNSUBSCRIBE, PUNSUBSCRIBE or SUNSUBSCRIBE request
+                // * `message` - a subscription message
+                //
+                // Additionally redis will:
+                // * accept a few regular commands while in pubsub mode: PING, RESET and QUIT
+                // * return an error response when a nonexistent or non pubsub compatible command is used
+                //
+                // Note: PING has a custom response when in pubsub mode.
+                //       It returns an array ['pong', $pingMessage] instead of directly returning $pingMessage.
+                //       But this doesnt cause any problems for us.
+
+                // Determine if message is a `message` subscription message
+                //
+                // Because PING, RESET, QUIT and error responses never return a RedisFrame::Array starting with `message`,
+                // they have no way to collide with the `message` value of a subscription message.
+                // So while we are in subscription mode we can use that to determine if an
+                // incoming message is a subscription message.
+                let is_subscription_message = if *is_subscribed {
+                    if let Some(Frame::Redis(RedisFrame::Array(array))) = message.frame() {
+                        if let [RedisFrame::BulkString(ty), ..] = array.as_slice() {
+                            ty.as_ref() == b"message"
+                        } else {
+                            false
                         }
+                    } else {
+                        false
                     }
+                } else {
+                    false
+                };
+
+                // Update is_subscribed state
+                //
+                // In order to make sense of a response we need the main task to
+                // send us the type of its corresponding request.
+                //
+                // In order to keep the incoming request MessageTypes in sync with their corresponding responses
+                // we must only process a MessageType when the message is not a subscription message.
+                // This is fine because subscription messages cannot affect the is_subscribed state.
+                if !is_subscription_message {
+                    match sent_message_type.recv().await {
+                        Some(MessageType::Subscribe) | Some(MessageType::Unsubscribe) => {
+                            if let Some(Frame::Redis(RedisFrame::Array(array))) = message.frame() {
+                                if let Some(RedisFrame::Integer(count)) = array.get(2) {
```

Comment:
Could you add a comment explaining that this `count` value is the number of channels we are currently subscribed to

Reply:
Is renaming the variable sufficient?

Reply:
Yeah or `subscribed_channel_count`. Not that familiar with the redis pub sub protocol so having that is helpful.